### PR TITLE
Enable Snowflake as a Database Resource

### PR DESCRIPTION
**Pull request by bjsemrad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

```release-note
To `vault_database_secret_backend_connection`, add support for the `snowflake-database-plugin`
```

Output from acceptance testing:

```
$make testacc TESTARGS='-run=TestAccDatabaseSecretBackendConnection_snowflake'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccDatabaseSecretBackendConnection_snowflake -timeout 120m
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/codegen   (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role        (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation      (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/util      (cached) [no tests to run]
=== RUN   TestAccDatabaseSecretBackendConnection_snowflake
hellow
--- PASS: TestAccDatabaseSecretBackendConnection_snowflake (2.80s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     2.847s
...
```

### CODE REVIEW

**Comment by bjsemrad** [comment]
I can remove the dependency updates, but I ran suggested vendor commands (and those were updated).

**Comment by robbruce** [comment]
@bjsemrad - can you get this rebased with master?

**Comment by bjsemrad** [comment]
@robbruce rebased, and updated mods/tidy sorry it took a bit.

**Comment by chriskuchin** [comment]
Would love to use this. Is there any plan to get this merged? Was just going to put this together basically the same and then found this one. Happy to look into the Conflicts if @bjsemrad is unable to at this point.

**Comment by bjsemrad** [comment]
@chriskuchin I’ll rebase this again tonight.

**Comment by chriskuchin** [comment]
No worries thanks appreciate that.

@robbruce once that is fixed is there anything else that is needed to get this merged?

**Comment by robbruce** [comment]
@chriskuchin - I’m not a code maintainer, just a consumer. Thanks for keeping it up to date though!

**Comment by chriskuchin** [comment]
ahh ok I will see if i can chase down looking at git history a maintainer after the conflicts are cleaned up and see if we can get this merged

**Comment by chriskuchin** [comment]
@tvoran you seem to be the only active maintainer on this repo. Once the conflicts are cleaned up do you need anything else to get this added to the provider? Happy to write up an issue or anything else you need to get this pulled into the provider.

Thanks

**Review by chriskuchin — APPROVED** [review]

**Comment by bjsemrad** [comment]
@chriskuchin PR was updated I'll try to watch this more closely.

**Comment by chriskuchin** [comment]
@bjsemrad awesome thanks.

@tvoran is there someone else I should be trying to coordinate this addition with? looking at the changelog and main branch commits seems like you are the only active maintainer. Happy to do whatever it takes to get this moved forward. Thanks!

File: .test-env [hunk 1]
```diff
@@ -4,3 +4,12 @@ export MYSQL_URL="vault:vault@tcp(mysql:3306)/"
...
 export ELASTIC_PASSWORD="elastic"
+export SNOWFLAKE_HOST="http://snowflake:443"
+export SNOWFLAKE_ACCOUNT="12345"
+export SNOWFLAKE_USERNAME="admin"
+export SNOWFLAKE_PASSWORD="admin"
+export SNOWFLAKE_REGION="us-east-1"
+export SNOWFLAKE_DB="snowdb"
+export SNOWFLAKE_SCHEMA="snowscheme"
+export SNOWFLAKE_WAREHOUSE="snowwarehouse"
+
```
Comment (tvoran) [inline .test-env:15]: I don't think .test-env is being used in CI anymore, so these can be removed.

```suggestion
```

File: vault/resource_database_secret_backend_connection.go [hunk 2]
```diff
@@ -273,6 +273,59 @@ func databaseSecretBackendConnectionResource() *schema.Resource {
...
+				Type:        schema.TypeList,
+				Optional:    true,
+				Description: "Connection parameters for the snowflake-database-plugin plugin.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"account": {
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "The Account Name for the snowflake connection. Formats <accountname> if in us-west-2, or <accountname>.<region> for any other region",
+						},
```
Comment (tvoran) [inline vault/resource_database_secret_backend_connection.go:286]: I'm curious why this schema breaks the `connection_url` API parameter up into its components? Is it just for convenience?

What do you think about just following the [API parameters](https://www.vaultproject.io/api/secret/databases/snowflake#parameters), and including good examples in the docs for different types of connection_url's? It also looks like in order to support root credential rotation, we need to be able to specify the `connection_url` separately from the `username` and `password`.

Also this schema seems to be missing the `max_open_connections`, `max_idle_connections`, `max_connection_lifetime`, and `username_template` parameters.

File: vault/resource_database_secret_backend_connection.go [hunk 3]
```diff
@@ -331,6 +340,28 @@ func mysqlConnectionStringResource() *schema.Resource {
...
+		Type:        schema.TypeString,
+		Required:    true,
+		Description: "The AccountAdmin level user using to connect to snowflake",
+	}
+	r.Schema["password"] = &schema.Schema{
+		Type:        schema.TypeString,
+		Required:    true,
+		Description: "The password with the provided user",
+		Sensitive:   true,
+	}
```
[the remark this answers is not in the document]
Comment (bjsemrad) [reply vault/resource_database_secret_backend_connection.go]: > Should username and password be marked as optional? I think they're only required if the connect_url is templated.

Yes they should I missed that, good catch.

File: website/docs/r/database_secret_backend_connection.md [hunk 4]
```diff
@@ -217,6 +219,30 @@ Exactly one of the nested blocks of configuration options must be supplied.
...
+* `max_idle_connections` - (Optional) The maximum number of idle connections to
+  maintain.
+
+* `max_connection_lifetime` - (Optional) The maximum number of seconds to keep
+  a connection alive for.
+
+* `username` - (Required) The username to be used in the connection (the account admin level).
+
+* `password` - (Required) The password to be used in the connection.
+*
```
Comment (tvoran) [inline website/docs/r/database_secret_backend_connection.md:241]:
```suggestion
* `username` - (Optional) The username to be used in the connection (the account admin level).

* `password` - (Optional) The password to be used in the connection.

```

File: website/docs/r/database_secret_backend_connection.md [hunk 5]
```diff
@@ -217,6 +219,30 @@ Exactly one of the nested blocks of configuration options must be supplied.
...
+  a connection alive for.
+
+* `username` - (Required) The username to be used in the connection (the account admin level).
+
+* `password` - (Required) The password to be used in the connection.
+*  
+* `username_template` - (Optional) - [Template](https://www.vaultproject.io/docs/concepts/username-templating) describing how dynamic usernames are generated.
+
+
+
```
Comment (tvoran) [inline website/docs/r/database_secret_backend_connection.md:245]: One blank line is sufficient here I think.

```suggestion
```

File: vault/resource_database_secret_backend_connection_test.go [hunk 6]
```diff
@@ -904,6 +935,28 @@ resource "vault_database_secret_backend_connection" "test" {
...
+  allowed_roles = ["dev", "prod"]
+  root_rotation_statements = ["FOOBAR"]
+
+  snowflake { 
+    connection_url = "%s"
+	username = "%s"
+	password = "%s"
+  }
+}
+`, path, name, url,  username, password)
```
Comment (tvoran) [inline vault/resource_database_secret_backend_connection_test.go:957]:
```suggestion
`, path, name, url, username, password)
```

**Review by tvoran — COMMENTED** [review]
Almost there! Just one suggestion in the test, and there are a couple spacing issues that `make fmt` will fix.

**Review by tvoran — APPROVED** [review]
Thanks!

**Comment by chriskuchin** [comment]
Thanks really appreciate all the work here!!!

**Pull request merged**